src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultGeneratedDocumentPublisher.cs

```diff
@@ -89,11 +129,19 @@ private void ProjectSnapshotManager_Changed(object sender, ProjectChangeEventArg
             {
                 case ProjectChangeKind.DocumentChanged:
                 case ProjectChangeKind.DocumentRemoved:
-                    if (_publishedSourceText.ContainsKey(args.DocumentFilePath) &&
+                    if (_publishedCSharpSourceText.ContainsKey(args.DocumentFilePath) &&
+                        !_projectSnapshotManager.IsDocumentOpen(args.DocumentFilePath))
+                    {
+                        // Document closed or removed, evict published source text.
+                        var removed = _publishedCSharpSourceText.Remove(args.DocumentFilePath);
+
+                        Debug.Assert(removed, "Published source text should be protected by the foreground thread and should never fail to remove.");
+                    }
+                    if (_publishedHtmlSourceText.ContainsKey(args.DocumentFilePath) &&
```

Comment:
Why have a separate check here. If the host document gets removed shouldn't we nuke both the CSharp & Html source texts?

Reply:
The separate checks is mainly because of the separate `ContainsKey` check in the two different dictionaries.

Reply:
oh derp derp 👍